modules/event-emitter/main/index.coffee

```diff
@@ -130,10 +134,22 @@ class EventEmitter
   # deregisters a callback
   off: (name, namespace, callback) ->
     if hx.isString(namespace)
-      @emittersMap.get(namespace)?.off(name, callback)
+      if @emittersMap.has(namespace)
+        be = @emittersMap.get(namespace)
+        be.off(name, callback)
+        if be.isEmpty()
+          removeEmitter(this, be, namespace)
     else
-      for emitter in @emitters.entries()
+      if not callback and not hx.isString(namespace)
+        callback = namespace
+
+      emitters = @emitters.entries()
+      emittersToRemove = []
+      emitters.forEach (emitter) =>
         emitter.off(name, callback)
+        if emitter isnt @global and emitter.isEmpty()
+          emittersToRemove.push(emitter)
+      emittersToRemove.map((e) => removeEmitter(this, e))
```

Comment:
This was originally using a for loop - I changed to `forEach` as it seems more reliable in this case.
I also split the removal out of the function as removing inside causes reference issues and 'leftover' emitters when doing `.off()` and using several namespaces.